docs/references/contributor_guide.md

```diff
@@ -1,22 +1,126 @@
+
+2. **Generate HTML Documentation**
+
+bash
+   make html
+
+   The generated static site can be found under the _build/html directory.
+
+3. **Local Preview**
+
+bash
+   bash serve.sh
+
+   Open your browser at the specified port to view the docs locally.
+
+4. **Clean Notebook Outputs & Build Artifacts**
+
+bash
+   find . -name '*.ipynb' -exec nbstripout {} \;
+   make clean
+
+   - **nbstripout** removes all outputs from Jupyter notebooks.
+   - **make clean** removes temporary files and _build artifacts.
+
+5. **Modify index.rst (If Necessary)**
+   Update index.rst whenever you add or remove sections from the documentation.
+
+6. **Pass Pre-Commit and Submit a PR**
+   After confirming the docs build successfully and pass all checks, open a Pull Request.
+
+CI will automatically build and validate the documentation to ensure consistency and completeness before merging.
+
+## 4. Running Unit Tests Locally & Adding to CI
+
+SGLang uses Python’s built-in [unittest](https://docs.python.org/3/library/unittest.html) framework. You can run tests either individually or in suites.
+
+### 4.1 Test Backend Runtime
+
+bash
+cd sglang/test/srt
+
+# Run a single file
+python3 test_srt_endpoint.py
+
+# Run a single test
+python3 -m unittest test_srt_endpoint.TestSRTEndpoint.test_simple_decode
+
+# Run a suite with multiple files
+python3 run_suite.py --suite minimal
+
+
+### 4.2 Test Frontend Language
+
+bash
+cd sglang/test/lang
+export OPENAI_API_KEY=sk-*****
+
+# Run a single file
+python3 test_openai_backend.py
+
+# Run a single test
+python3 -m unittest test_openai_backend.TestOpenAIBackend.test_few_shot_qa
+
+# Run a suite with multiple files
+python3 run_suite.py --suite minimal
+
+
+### 4.3 Adding or Updating Tests in CI
+
+- Create new test files under test/srt, ensure they are referenced in test/srt/run_suite.py, or they will not be picked up.
+- In CI, all tests are executed automatically. You may modify the corresponding workflow in .github/workflows/ if you wish to include custom test groups or additional checks.
+
+### 4.4 Writing Elegant Test Cases
+
+- Examine existing tests in [sglang/test](https://github.com/sgl-project/sglang/tree/main/test) for practical examples.
+- Keep each test function focused on a specific scenario or unit of functionality.
+- Give tests descriptive names reflecting their purpose.
+- Use robust assertions (e.g., assert, unittest methods) to validate outcomes.
+- Clean up resources to avoid side effects and preserve test independence.
+
+## 5. Tips for Newcomers
+
+If you want to contribute but don’t have a specific idea in mind, start by picking up issues labeled [“good first issue” or “help wanted”](https://github.com/sgl-project/sglang/issues?q=is%3Aissue+label%3A%22good+first+issue%22%2C%22help+wanted%22). These tasks typically have lower complexity and provide an excellent introduction to the codebase. Feel free to ask questions or request clarifications in the corresponding issue threads.
```

Comment:
Introduce this:

https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/tree/main/sglang/code-walk-through